receiver/prometheusexec/subprocessmanager/manager.go

```diff
@@ -0,0 +1,150 @@
+
+package subprocessmanager
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math"
+	"math/rand"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/kballard/go-shellquote"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusexec/subprocessmanager/config"
+)
+
+// Process struct holds all the info needed for subprocesses
+type Process struct {
+	Command    string
+	Port       int
+	Env        []config.EnvConfig
+	CustomName string
+}
+
+const (
+	// HealthyProcessTime is the default time a process needs to stay alive to be considered healthy
+	HealthyProcessTime time.Duration = 30 * time.Minute
+	// healthyCrashCount is the amount of times a process can crash (within the healthyProcessTime) before being considered unstable - it may be trying to find a port
+	HealthyCrashCount int = 3
+	// delayMutiplier is the factor by which the delay scales (default is doubling every crash)
+	delayMultiplier float64 = 2.0
+	// baseDelay is the base exponential backoff delay that every process waits for before restarting
+	baseDelay time.Duration = 1 * time.Second
+)
+
+// Run will start the process and keep track of running time
+func (proc *Process) Run(logger *zap.Logger) (time.Duration, error) {
+	var (
+		start     time.Time
+		elapsed   time.Duration
+		argsSlice []string
+	)
+
+	// Parse the command line string into arguments
+	args, err := shellquote.Split(proc.Command)
+	if err != nil {
+		return elapsed, fmt.Errorf("[%v] could not parse command error: %v", proc.CustomName, err)
+	}
+	// Separate the executable from the flags for the Command object
+	if len(args) > 1 {
+		argsSlice = args[1:]
+	}
+
+	// Create the command object and attach current os environment + environment variables defined by user
+	childProcess := exec.Command(args[0], argsSlice...)
+	childProcess.Env = os.Environ()
+	childProcess.Env = append(childProcess.Env, formatEnvSlice(&proc.Env)...)
+
+	// Handle the subprocess output in a goroutine, and pipe the stderr/stdout into one stream
+	cmdReader, err := childProcess.StdoutPipe()
+	if err != nil {
+		return elapsed, fmt.Errorf("[%v] could not get the command's output pipe, err: %v", proc.CustomName, err)
+	}
+	childProcess.Stderr = childProcess.Stdout
+	go proc.pipeSubprocessOutput(bufio.NewReader(cmdReader), logger)
+
+	// Start and stop timer (elapsed) right before and after executing the command
+	start = time.Now()
+	errProcess := childProcess.Start()
+	if errProcess != nil {
+		return elapsed, fmt.Errorf("[%v] process could not start: %v", proc.CustomName, errProcess)
+	}
+
+	errProcess = childProcess.Wait()
+	elapsed = time.Since(start)
+	if errProcess != nil {
+		return elapsed, fmt.Errorf("[%v] process error: %v", proc.CustomName, errProcess)
+	}
+
+	return elapsed, nil
+}
+
+// Log every line of the subprocesse's output using zap
+func (proc *Process) pipeSubprocessOutput(reader *bufio.Reader, logger *zap.Logger) {
+	var (
+		line string
+		err  error
+	)
```

Comment:
I don't _think_ this is common in go. Normally I would just see variables intialized when first used, e.g. `line, err := reader.ReadString('\n')`

Reply:
I was thinking that since this loop might run a lot to not re-declare the variable every time for performance reasons, but I'll change it for convention! Thanks

Reply:
I'm pretty sure the compiler will do that kind of optimisation for you